File: src/stat_node.cpp
#include <deque>
#include <numeric>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "a2_controller/srv/get_velocity_stats.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class StatsNode : public rclcpp::Node {
// constructor
public:
    StatsNode() : Node("stats_node") {
        // listens to what the control node is pushing (for the movement)
        sub_vel_ = this->create_subscription<geometry_msgs::msg::Twist>("/cmd_vel", 10, std::bind(&StatsNode::vel_callback, this, _1));

        // creates an endpoint and waits for user requests. when it gets it will run the function get_...
        srv_stats_ = this->create_service<a2_controller::srv::GetVelocityStats>("get_velocity_stats", std::bind(&StatsNode::get_stats_callback, this, _1, _2));
    }



private:

    // implement a fifo queue
    void vel_callback(const geometry_msgs::msg::Twist::SharedPtr msg) {
        // we store only 5 values. if more than 5 we pop
        if (history_.size() >= 5) {
            history_.pop_front();
        }
        history_.push_back(msg->linear.x);

        // same for angular
        if(history_a.size() >= 5){
            history_a.pop_front();
        }
        history_a.push_back(msg->angular.z);    // it is z bc we are calc the angular velocity wrt to the z axis
    }


    void get_stats_callback(
        const std::shared_ptr<a2_controller::srv::GetVelocityStats::Request>,std::shared_ptr<a2_controller::srv::GetVelocityStats::Response> res) 
    {
        if (history_.empty()) {
            res->avg_linear_vel = 0.0;
        }
        else{
            // calc the avg for linear
            float sum = std::accumulate(history_.begin(), history_.end(), 0.0);
            res->avg_linear_vel = sum / history_.size();
        }

        if(history_a.empty()){
            res->avg_angular_vel = 0.0;
            return;
        }
        else{
              // angular
            float sum_a = std::accumulate(history_a.begin(), history_a.end(), 0.0);
            res->avg_angular_vel = sum_a / history_a.size(); 
        }
        
        RCLCPP_INFO(this->get_logger(), "Stats - Linear: %.2f, Angular: %.2f", res->avg_linear_vel, res->avg_angular_vel);
    }

    // variables
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_vel_;
    rclcpp::Service<a2_controller::srv::GetVelocityStats>::SharedPtr srv_stats_;
    std::deque<float> history_;
    std::deque<float> history_a;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<StatsNode>());
    rclcpp::shutdown();
    return 0;
}

File: a2_python/a2_python/stat_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from a2_interfaces_python.srv import GetVelocityStats
from collections import deque

class StatsNode(Node):

    def __init__(self):
        super().__init__('stats_node')

        self.history_linear = deque(maxlen=5)
        self.history_angular = deque(maxlen=5)

        self.create_subscription(Twist, '/cmd_vel', self.vel_callback, 10)
        self.create_service(GetVelocityStats, 'get_velocity_stats', self.get_stats_callback)

        self.get_logger().info("The stats node has started")

    def vel_callback(self, msg):
        self.history_linear.append(msg.linear.x)
        self.history_angular.append(msg.angular.z)

    def get_stats_callback(self, request, response):
        # --- Linear ---
        if len(self.history_linear) == 0:
            response.avg_linear_vel = 0.0
        else:
            avg_lin = sum(self.history_linear) / len(self.history_linear)
            response.avg_linear_vel = float(avg_lin)
        
        # --- Angular ---
        if len(self.history_angular) == 0:
            response.avg_angular_vel = 0.0
        else:
            avg_ang = sum(self.history_angular) / len(self.history_angular)
            # FIX: You were overwriting linear_vel here!
            response.avg_angular_vel = float(avg_ang) 

        self.get_logger().info(
            f"Stats - Linear: {response.avg_linear_vel:.2f}, Angular: {response.avg_angular_vel:.2f}"
        )

        return response

def main(args=None):
    rclpy.init(args=args)
    node = StatsNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()